include/deal.II/lac/vector_space_vector.h

```diff
@@ -0,0 +1,181 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2015 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__vector_space_vector_h
+#define dealii__vector_space_vector_h
+
+DEAL_II_NAMESPACE_OPEN
+
+#include <deal.II/base/mpi.h>
+#include <deal.II/base/numbers.h>
+
+namespace LinearAlgebra
+{
+  template <typename Number>
+  class ReadWriteVector<Number>;
+
+  class IndexSet;
+
+  /*! @addtogroup Vectors
+   *@{
+   */
+
+  /**
+   * VectorSpaceVector is an abstract class that is used to define the interface
+   * that vector classes need to implement when they want to implement global
+   * operations. This class is complementary of ReadWriteVector which allows
+   * the access of individual elements but does not allow global operations.
+   *
+   * @author Bruno Turcksin, 2015.
+   */
+  template <Number>
+  class VectorSpaceVector
+  {
+  public:
+    typedef types::global_dof_index                           size_type;
+    typedef typename numbers::NumberTraits<Number>::real_type real_type;
+
+    /**
+     * Imports all the elements present in the vector's IndexSet from the input
+     * vector @p read_write_vector.
+     */
+    virtual VectorSpaceVector<Number> &operator= (const ReadWriteVector<Number> &read_write_vector) = 0;
```

Comment:
Is this the only way to go from a read-write vector to a vector space vector?

I am thinking about the general workflow in a time-dependent problem and how I should use read-write versus vector-space vectors. This interface means that I will mostly have vector-space vectors to represent all my solution vectors. This way, I can add vectors by add/sadd/equ and do the usual stuff on vectors. When I go to an assembly routine which involves old time step data, I will copy the data to a readwrite vector and do whatever is necessary. I will have another read-write vector to collect my residual. In the end of assembly, I will call operator= on a vector-space vector to gather all data. So doing the time steps look fine. If efficiency is important, one might not actually copy the data but rather re-use some pointers in a suitable way.

However, what do we do with the function interpolating initial conditions? For that, I need to access individual vector elements, i.e., a ReadWriteVector. However, I need to define a different 'combine mode' here than for the assembly ('insert' rather than 'add') because otherwise I will duplicate some vector entries. (Or is the DoFHandler supposed to only insert locally owned data? That calls for errors.)

Thus, I see the need for a VectorOperation argument here.